### PR TITLE
improve recovery displays workflow

### PR DESCRIPTION
**Pull request by Robertgervais**

### CODE REVIEW

**Comment by Robertgervais** [comment]
https://github.com/bikeindex/bike_index/blob/4865f5635b56d6fefcf6bdaa11e86c76430728c2/spec/models/stolen_record_spec.rb

File: app/controllers/bikes/recovery_controller.rb [hunk 1]
```diff
@@ -12,7 +12,9 @@ def edit
     end
 
     def update
-      if @stolen_record.add_recovery_information(permitted_params)
+      update_params = permitted_params.merge(recovering_user: current_user)
```
Comment (sethherr) [inline app/controllers/bikes/recovery_controller.rb]: Use the way this is done on master - it's in the permitted params.

File: app/controllers/bikes/recovery_controller.rb [hunk 2]
```diff
@@ -12,7 +12,9 @@ def edit
...
+      if @stolen_record.add_recovery_information(update_params)
+        @stolen_record.update_attributes(recovery_display_status: 3)
```
Comment (sethherr) [inline app/controllers/bikes/recovery_controller.rb]: Should be a string

File: app/controllers/bikes/recovery_controller.rb (hunk 1 above)
Comment (sethherr) [reply app/controllers/bikes/recovery_controller.rb]: View the changes done here: https://github.com/bikeindex/bike_index/pull/846/files

File: app/models/stolen_record.rb [hunk 3]
```diff
@@ -168,14 +185,26 @@ def tsv_row(with_article = true, with_stolen_locations: false)
     row
   end
 
+  def calculated_recovery_display_status
+    return "not_eligible" unless can_share_recovery
+    return "not_displayed" if not_displayed?
+    return "displayed" if displayed?
```
Comment (sethherr) [inline app/models/stolen_record.rb]: Are you going to set this to `displayed` somewhere else in the future?

File: app/models/stolen_record.rb [hunk 4]
```diff
@@ -168,14 +185,26 @@ def tsv_row(with_article = true, with_stolen_locations: false)
...
   def add_recovery_information(info = {})
     info = ActiveSupport::HashWithIndifferentAccess.new(info)
     self.date_recovered = TimeParser.parse(info[:date_recovered], info[:timezone]) || Time.now
     update_attributes(current: false,
                       recovered_description: info[:recovered_description],
                       recovering_user_id: info[:recovering_user_id],
                       index_helped_recovery: ("#{info[:index_helped_recovery]}" =~ /t|1/i).present?,
-                      can_share_recovery: ("#{info[:can_share_recovery]}" =~ /t|1/i).present?)
+                      can_share_recovery: ("#{info[:can_share_recovery]}" =~ /t|1/i).present?,
+                      recovery_display_status: calculated_recovery_display_status)
```
Comment (sethherr) [inline app/models/stolen_record.rb]: you don't need to do this, it happens automatically in the callback

File: app/models/stolen_record.rb (hunk 3 above)
Comment (Robertgervais) [reply app/models/stolen_record.rb]: No, the only place it gets set to displayed, currently, is when the display get's created.

File: app/models/stolen_record.rb [hunk 5]
```diff
@@ -100,7 +109,15 @@ def self.locking_defeat_description_select
     locking_defeat_description.map { |l| [l, l] }
   end
 
-  before_save :set_phone, :fix_date, :titleize_city, :update_tsved_at
+  before_save :set_calculated_attributes
```
Comment (sethherr) [inline app/models/stolen_record.rb]: The new approach I'm taking with putting all callbacks (ie `before_save ...`) up at the top, after the scopes. It makes it easier to find. You don't need to move the method, but I would like it if you moved this up there to the top!

File: app/models/stolen_record.rb (hunk 3 above)
Comment (sethherr) [reply app/models/stolen_record.rb]: Great! That makes sense.

**Comment by Robertgervais** [comment]
Responded. I took `recovery_display_status` out of `add_recovery_info`, and all of THE TESTS I WROTE are still green, but it still won't let me set enum status in console when I remove this. I know what before_save is supposed to do, and it makes sense not to have it in two places, but without it I can't seem to get the `waiting_on_decision` status set and I'm not sure why.

**Comment by Robertgervais** [comment]
#739

**Pull request merged**